src/App.tsx

```diff
@@ -1,24 +1,228 @@
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [title, setTitle] = useState('');
+  const [isInputDisabled, setIsInputDisabled] = useState(false);
+  const [temporaryTodo, setTemporaryTodo] = useState<Todo>();
+  const [isLoading, setIsLoading] = useState(true);
+  const [loadingIds, setLoadingIds] = useState<number[]>([]);
+  const [selectedFilter, setSelectedFilter] = useState(FilterType.All);
+
+  useEffect(() => {
+    const getTodosFromServer = async () => {
+      try {
+        setErrorMessage('');
+        const todosFromServer = await getTodos(USER_ID);
+
+        setTodos(todosFromServer);
+      } catch (error) {
+        if (error instanceof Error) {
+          setErrorMessage(error.message);
+        }
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    getTodosFromServer();
+  }, []);
+
+  const addTodo = (todoTitle: string) => {
+    const newTodo = {
+      title: todoTitle,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setIsInputDisabled(true);
+
+    setTemporaryTodo({ ...newTodo, id: 0 });
+
+    postTodo(newTodo)
+      .then(result => {
+        setTodos(state => [...state, result]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+      })
+      .finally(() => {
+        setIsInputDisabled(false);
+        setTemporaryTodo(undefined);
+      });
+  };
+
+  const handleFormSubmit = useCallback((event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (!title) {
+      setErrorMessage('Tittle cannot be empty');
+      setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+
+      return;
+    }
+
+    addTodo(title);
+    setTitle('');
+  }, [title]);
+
+  const removeTodo = useCallback(async (id: number) => {
+    setLoadingIds(state => [...state, id]);
+    try {
+      await deleteTodo(id);
+      setTodos(todos.filter(todo => todo.id !== id));
+    } catch {
+      setErrorMessage('Unable to delete a todo');
+      setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+    } finally {
+      setLoadingIds(state_1 => state_1.filter(el => el !== id));
+    }
+  }, [todos]);
+
+  const activeTodos = useMemo(() => {
+    return todos.filter(({ completed }) => !completed).length;
+  }, [todos]);
+
+  const handleInputChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setTitle(event.target.value);
+    }, [title],
+  );
+
+  const removeError = () => {
+    setErrorMessage('');
+  };
+
+  const handleUpdate = async (id: number, data: Partial<Todo>) => {
+    setLoadingIds(state => [...state, id]);
+
+    try {
+      await patchTodo(id, data);
+
+      setTodos(state => state.map(todo => {
+        if (todo.id === id) {
+          return { ...todo, ...data };
+        }
+
+        return todo;
+      }));
```

Comment:
It's a bad idea to use local data after update to set it to state
Data can be changed on server side or because of some error it can be not processed etc.
Use data from API as source of truth, `patch` returns updated data, use it instead of local